doc/source/development.rst

```diff
@@ -54,7 +54,15 @@ helpful.
    something like ``flake8 ray/python/ray/worker.py``. You may need to first run
    ``pip install flake8``.
 
-5. **Inspecting Redis shards by hand:** To inspect the primary Redis shard by
+5. **Autoformatting code**. We use ``yapf``
+   https://github.com/google/yapf for linting, and the config file is
+   located at ``.style.yapf``. We recommend adding ``.travis/yapf.sh``
+   to git's ``pre-push`` hook, so that you never have to worry about
+   your code formatting ever again. Note that RLlib is not currently
+   autoformatted, though running ``yapf`` over the individual files you
+   edit with ``yapf -i -p -r <file1> <file2> <...>`` is a good idea.
```

Comment:
How about this instead?
```
yapf -i -p `git diff --name-only upstream/master | grep .py`
```

Reply:
Not sure it's actually a good idea to run `yapf` on the files you edit because that will introduce a lot of changes in places you didn't mean to change and will make it hard to see what your PR is doing.

Reply:
Hm I see, perhaps the --exclude flags would work here? Though then the cmd line becomes a bit long so you probably want to check in the script.

It would be great to have a ./yapf script that formats all relevant files and also is fast enough to run all the time.

Reply:
This problem is actually why I've been pushing so annoyingly hard for autoformat everywhere, all the time. If we bit the bullet, formatted all the python code and read the ugly PR, and then put a pre-*commit* hook (we could copy [this script from go](https://tip.golang.org/misc/git/pre-commit)), no change in the codebase would ever introduce line noise, because it would always be formatted the same. Since we'd only ever format changed files, it'd be fast enough to be unnoticeable too. It could later open up the possibility of [mechanical source transformation](https://blog.golang.org/go-fmt-your-code).

A pre-push hook is a halfway measure, but is still far better than what's currently done. It also saves having to write overly complex scripts for what's really a problem with standards. 

30k lines of ray were autoformatted without any major issues. I think the rest of the python code could be done too. I'd be happy to review, since code formatting hardly requires deep knowledge of the actual code being run.

Any PRs against `master` could run yapf in their final commit and then only the diff would appear in the github PR review tool, so it wouldn't be a pain to review.

Reply:
You can still add a pre-commit hook, just with --excludes for now, no?

FWIW I think the full linting will happen, it's just that we need to find some quiescent time for the sub-projects and then someone can do a quick 3am push to reformat everything.